src/OmniSharp.Roslyn.CSharp/Services/Refactoring/RunFixAllCodeActionService.cs

```diff
@@ -44,76 +46,140 @@ public class RunFixAllCodeActionService : BaseCodeActionService<RunFixAllRequest
+                if (document is null)
                 {
-                    _logger.LogError($"Running fix all action {singleFixableProviderWithDocument} in document {singleFixableProviderWithDocument.DocumentPath} prevented by error: {ex}");
+                    throw new InvalidOperationException("Could not find the document with the diagnostic in the solution");
                 }
             }
 
-            var changes = await GetFileChangesAsync(Workspace.CurrentSolution, solutionBeforeChanges, Path.GetDirectoryName(request.FileName), request.WantsTextChanges, request.WantsAllCodeActionOperations);
+            _logger.LogTrace("{0} is still present in the document. Getting fixes.", diagnosticId);
+            CodeAction action = null;
+            var context = new CodeFixContext(document, primaryDiagnostic,
+                (a, _) =>
+                {
+                    if (action == null)
+                    {
+                        action = a;
+                    }
+                },
+                cancellationToken);
+
+            await codeFixProvider.RegisterCodeFixesAsync(context).ConfigureAwait(false);
 
-            return new RunFixAllResponse
+            var roslynScope = scope switch
             {
-                Changes = changes.FileChanges
+                FixAllScope.Document => RoslynFixAllScope.Document,
+                FixAllScope.Project => RoslynFixAllScope.Project,
+                FixAllScope.Solution => RoslynFixAllScope.Solution,
+                _ => throw new InvalidOperationException()
             };
+
+            var fixAllContext = new FixAllContext(document, codeFixProvider, roslynScope, action.EquivalenceKey, ImmutableArray.Create(diagnosticId), _fixAllDiagnosticProvider, cancellationToken);
+
+            _logger.LogTrace("Finding FixAll fix for {0}.", diagnosticId);
+            var fixes = await fixAllProvider.GetFixAsync(fixAllContext);
+            if (fixes == null)
+            {
+                _logger.LogInformation("FixAll not found for {0}.", diagnosticId);
+                return originalDoc;
+            }
+
+            _logger.LogTrace("Getting FixAll operations for {0}.", diagnosticId);
+            var operations = await fixes.GetOperationsAsync(cancellationToken);
+
+            // Currently, there are no roslyn changes that will result in multiple ApplyChangesOperations
+            Debug.Assert(operations.OfType<ApplyChangesOperation>().Count() < 2);
+            if (operations.OfType<ApplyChangesOperation>().FirstOrDefault() is ApplyChangesOperation applyChangesOperation)
+            {
+                _logger.LogTrace("Found apply changes operation for {0}.", diagnosticId);
+                return applyChangesOperation.ChangedSolution.GetDocument(originalDoc.Id);
+            }
+            else
+            {
+                _logger.LogTrace("No apply changes operation for {0}.", diagnosticId);
```

Comment:
there will be some cases when the code action did not produce `ApplyChangesOperation` (e.g. modifying files that are not part of the solution)

currently in the `RunCodeActionService` we'd try to apply those directly (https://github.com/OmniSharp/omnisharp-roslyn/blob/master/src/OmniSharp.Roslyn.CSharp/Services/Refactoring/V2/RunCodeActionService.cs#L73-L74). I think it would be consistent to have same behavior here

Reply:
I agree, but I think that's outside the scope of this PR. I'd be happy to file an issue for following up on this but I'd prefer this change not have any more changes then necessary to decouple the service from the workspace.